netshare/utils/field.py

```diff
@@ -1,73 +1,124 @@
+import json
+from typing import List, Optional, Tuple, Union
+
 import numpy as np
 import pandas as pd
-from .output import Normalization, OutputType, Output
+from annoy import AnnoyIndex
+from config_io import Config
+
+from netshare.learn.utils.word2vec_embedding import (
+    get_original_objs,
+    get_vector,
+    get_word2vec_type_col,
+)
+from netshare.utils.paths import (
+    get_annoy_dict_idx_ele_for_word2vec,
+    get_annoyIndex_for_word2vec,
+)
+
+from .output import Normalization, Output, OutputType
 
 EPS = 1e-8
+FieldKey = Tuple[str, str, str]
 
 
 class Field(object):
-    def __init__(self, name):
+    def __init__(self, name: Union[str, List[str]], log1p_norm: bool = False):
         self.name = name
+        self.log1p_norm = log1p_norm
+
+    def normalize(self, x):
+        raise NotImplementedError
 
-    def normalize(self):
+    def denormalize(self, x):
         raise NotImplementedError
 
-    def denormalize(self):
+    def get_output_type(self):
         raise NotImplementedError
 
-    def getOutputType(self):
+    def get_output_dim(self) -> int:
         raise NotImplementedError
 
 
 class ContinuousField(Field):
-    def __init__(self, norm_option, min_x=None, max_x=None, dim_x=1,
-                 *args, **kwargs):
-        super(ContinuousField, self).__init__(*args, **kwargs)
+    def __init__(
+        self,
+        name: Union[str, List[str]],
+        norm_option: Normalization,
+        min_x: float,
+        max_x: float,
+        dim_x: int = 1,
+        log1p_norm: bool = False,
+    ) -> None:
+        super(ContinuousField, self).__init__(name=name, log1p_norm=log1p_norm)
 
         self.min_x = min_x
         self.max_x = max_x
         self.norm_option = norm_option
         self.dim_x = dim_x
+        if self.log1p_norm:
+            # If set log1p_norm, we need to re-calculate the min_x and max_x after doing
+            # log1p. Since the denormalizing step would first do min-max denomalization
+            # and then expm1. Using stale min_x and max_x would cause the synthetic
+            # value larger than expected and is possible to have np.inf
+            self.min_x = np.log1p(self.min_x)
+            self.max_x = np.log1p(self.max_x)
 
     # Normalize x in [a, b]: x' = (b-a)(x-min x)/(max x - minx) + a
     def normalize(self, x):
         if x.shape[-1] != self.dim_x:
-            raise ValueError(f"Dimension is {x.shape[-1]}. "
-                             f"Expected dimension is {self.dim_x}")
+            raise ValueError(
+                f"Dimension is {x.shape[-1]}. Expected dimension is {self.dim_x}"
+            )
+        if self.log1p_norm:
+            x = np.log1p(x)
+
         # [0, 1] normalization
         if self.norm_option == Normalization.ZERO_ONE:
+            if self.max_x - self.min_x == 0:
+                raise Exception("Not enough data to proceed!")
```

Comment:
I think this error is due to "constant" values instead of "not enough data".

Constant values happen a lot and why not just add a small epsilon (e.g., 1e-8) to `self.max_x - self.min_x`?

Reply:
Sounds absolutely correct!